src/ast/passes/codegen_llvm.cpp

```diff
@@ -2280,23 +2280,47 @@ void CodegenLLVM::createTupleCopy(const SizedType &expr_type,
 
 void CodegenLLVM::visit(Tuple &tuple)
 {
-  // Store elements on stack
   llvm::Type *tuple_ty = b_.GetType(tuple.type);
 
   compareStructure(tuple.type, tuple_ty);
 
-  std::vector<std::pair<llvm::Value *, const location *>> vals;
   std::vector<ScopedExprDeleter> scoped_dels;
-  vals.reserve(tuple.elems.size());
 
-  for (Expression *elem : tuple.elems) {
+  auto buf = b_.CreateTupleScratchBuffer(tuple.loc, tuple_depth_);
+  size_t tuple_size = datalayout().getTypeAllocSize(tuple_ty);
+  b_.CreateMemsetBPF(buf, b_.getInt8(0), tuple_size);
+
+  tuple_depth_++;
+  for (size_t i = 0; i < tuple.elems.size(); i++) {
+    // Evaluate expression using DFS
+    auto elem = tuple.elems[i];
     scoped_dels.emplace_back(accept(elem));
-    vals.push_back({ expr_, &elem->loc });
+
+    // Then copy or store child expression
+    // For child tuples, to avoid the child overwriting the parent's tuple
+    // scratch map, we use a separate scratch map key according to the tuple
+    // depth e.g. child key = N + 1 and parent key = N
```

Comment:
Feel free to ignore this but it took me a second to understand what this was saying. Maybe slight rewording:

```suggestion
    // For child tuples, to avoid the child overwriting the parent's tuple
    // scratch map, we use tuple_depth_ as the scratch map key 
    // e.g. for ("a", ("b", ("c", "d"))) the top tuple key (containing "a") is 0, then the inner tuple key (containing "b") is 1, then the tuple containing "c" and "d", it's key is 2 and so forth.
```

Reply:
With my new update we're no longer using this approach :) So resolving but lets re-review the new version and see where I can add comments!